File: leetcode/340. Longest Substring with At Most K Distinct Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        unordered_map<char, int> m;
        int max_len = 0;
        int i = 0;
        
        for(int j = 0; j < s.size(); ++j){
            if(m.find(s[j]) != m.end()){
                m[s[j]] += 1;
                max_len = max(max_len, j - i + 1);
            } else {
                m[s[j]] = 1;
                if(m.size() <= k){
                    max_len = max(max_len, j - i + 1);
                } else {
                    while(m.size() > k){
                        --m[s[i]];
                        if(m[s[i]] == 0){
                            m.erase(s[i]);
                        }
                        ++i;
                    }
                }
            }
        }
        
        return max_len;
    }
};

File: leetcode/340. Longest Substring with At Most K Distinct Characters.py
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        dic = {}
        max_len = 0
        i = 0
        
        for j, c in enumerate(s):
            if c in dic:
                dic[c] += 1
                max_len = max(max_len, j - i + 1)
            else:
                dic[c] = 1
                if len(dic) <= k:
                    max_len = max(max_len, j - i + 1)
                else:
                    while i <= j:
                        dic[s[i]] -= 1
                        if dic[s[i]] == 0:
                            del dic[s[i]]
                            i += 1
                            break
                        i += 1
        return max_len
            
